### ISSUE TITLE
bug in `grouped_` variants of functions?

### ISSUE THREAD

**Issue by IndrajeetPatil** [IndrajeetPatil/ggstatsplot#84]
As the `README` mentions, all functions in `ggstatsplot` are supposed to work irrespective of whether you enter a character (`x = "x"`) or or a bare expression (`x = x`). But this doesn't seem to be working for `grouped_` variants of functions for the `grouping.var` argument? 

Definintely something is not right with the way I've implemented this using `rlang`, my Achilles heel.

@ibecav You wanna take a look at this?

``` r
library(ggstatsplot)

# works
ggstatsplot::grouped_ggbetweenstats(
  data = dplyr::sample_frac(tbl = ggstatsplot::movies_long, size = 0.25) %>%
    dplyr::filter(.data = ., mpaa %in% c("R", "PG-13"), genre %in% c("Drama", "Comedy")),
  x = genre,
  y = rating,
  grouping.var = mpaa
)
```

![](https://i.imgur.com/Om99dpy.png)

``` r

# doesn't work
ggstatsplot::grouped_ggbetweenstats(
  data = dplyr::sample_frac(tbl = ggstatsplot::movies_long, size = 0.25) %>%
    dplyr::filter(.data = ., mpaa %in% c("R", "PG-13"), genre %in% c("Drama", "Comedy")),
  x = genre,
  y = rating,
  grouping.var = "mpaa"
)
#> Error in arrange_impl(.data, dots): incorrect size (1) at position 1, expecting : 341
```

<sup>Created on 2018-11-13 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by ibecav** [comment]
Yeah I'll take a look tomorrow.  To be honest quoting and unquoting makes my head hurt too.

**Comment by ibecav** [comment]
Took a quick look the bug starts with this line of code...

title.text = !!rlang::enquo(grouping.var) haven't quite figured out what to do about it yet.  I'll look tomorrow

**Comment by ibecav** [comment]
About to submit a pull request that fixes the immediate problem we should probably add a couple more lines as well.  Although the fix was simple tracing down was tedious and time consuming and i had never heard of `ensym` before

``` r
library(ggstatsplot)
# works now
ggstatsplot::grouped_ggbetweenstats(
  data = dplyr::sample_frac(tbl = ggstatsplot::movies_long, size = 0.25) %>%
  dplyr::filter(.data = ., mpaa %in% c("R", "PG-13"), genre %in% c("Drama", "Comedy")),
  x = genre,
  y = rating,
  grouping.var = "mpaa"
)
#> Note: Shapiro-Wilk Normality Test for rating : p-value = 0.502
#> 
#> Note: Bartlett's test for homogeneity of variances for factor genre : p-value = 0.437
#> 
#> Note: Shapiro-Wilk Normality Test for rating : p-value = 0.001
#> 
#> Note: Bartlett's test for homogeneity of variances for factor genre : p-value = 0.082
#> 
#> Warning: Plots from `grouped_` functions are not `ggplot` objects 
#> and can't be further modified with `ggplot2` functions.
#> 
```

![](https://i.imgur.com/mfyCDn5.png)

<sup>Created on 2018-11-14 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by ibecav** [comment]
Why don't you verify this initial PR and accept it as appropriate then I'll spend some time confirming that ...

1.  adding the same mechanism for other variables, in this case `x` & `y` is appropriate ( istill don't actually know why it is necessary for `grouping.variable` but not x & y

2.  I'll look at the other `grouped_` functions and confirm that this fix works for them as well.

By the way https://adv-r.hadley.nz/quasiquotation.html was the real hero here...

**Comment by IndrajeetPatil** [comment]
Cool, just checked that this works for `ggscatterstats`. 

Yeah, even I am stumped by why this needs to be the case only for `grouping.var`, but not for `x` and `y`. The latter rely on `!!rlang::enquo()` and work with both character and bare expression inputs but the same doesn't work for `grouping.var`, which is puzzling. I will check out Hadley's chapter.

**Comment by IndrajeetPatil** [comment]
Just pushed the changed version of `grouped_ggscatterstats` to `master`. I think it's going to be imperative to also include tests to check that both bare expressions and characters are working. I have added such tests for `grouped_ggbetweenstats` and `grouped_ggscatterstats`:
https://github.com/IndrajeetPatil/ggstatsplot/commit/90b8f4e9cf06a6f0d40501a885882dc7845075af

**Comment by IndrajeetPatil** [comment]
Thanks for getting to the bottom of this, Chuck! I concur with your recommendations about prefiltering. Let's implement this! Lemme know if you wanna do a PR or if I can take care of this and you work on tests. Either is fine with me.

Since we are on the topic of `rlang`, I am wondering if you can also think of a solution to sidestep this ugly hack I have implemented for `grouped_` functions when one of the variables is `NULL`.
e.g., https://github.com/IndrajeetPatil/ggstatsplot/blob/90b8f4e9cf06a6f0d40501a885882dc7845075af/R/grouped_ggbetweenstats.R#L150-L270
or 
https://github.com/IndrajeetPatil/ggstatsplot/blob/90b8f4e9cf06a6f0d40501a885882dc7845075af/R/grouped_ggscatterstats.R#L207-L380
and so on.

Copy-pasting code like this is a disaster waiting to happen and I haven't had a chance to dive into this and refactor the code to avoid repitition since I've been prioritizing adding new functionality for the next release. Wanna give this a shot as well?

**Comment by ibecav** [comment]
Well since you know all the places that little snippet has to go and can implement directly please go right ahead.

I'll take a look at this ugly issue and/or go back to writing tests...

**Comment by IndrajeetPatil** [comment]
Cool, I will start working on this.

**Comment by ibecav** [comment]
For what it's worth spent quite a few hours on the grouped_ to purrr branching you're doing.  No luck yet.  That's why I stopped banging my head on the wall and wrote the two tests you asked for.
